woocommerce/woocommerce-gateway-stripe
# SCA: Authenticate if necessary when changing subscription payment method

Fixes https://github.com/woocommerce/woocommerce-gateway-stripe/issues/1046

##  Changes proposed

Upon updating the payment method for a subscription, create a SetupIntent, and depending on its status, **redirect to an (SCA) authentication screen** (of the exact sort already in place for the "Pay for order" page) before proceeding with the redirect 👍

![Screenshot_2020-12-09 Checkout](https://user-images.githubusercontent.com/1867547/101696215-c6f38500-3a43-11eb-84a4-382ea520c01e.png)

Things to note with this implementation:

- The **source is still updated on the subscription when the form is initially submitted**, not after authentication succeeds. This _could_ be avoided by using a filter, but the _notice_ ("Payment method updated.") would still show on the authentication screen page load – see [source](https://github.com/woocommerce/woocommerce-subscriptions/blob/7df33fbbcc089c3d0d7ea0f64765ddfd41dca09e/includes/class-wc-subscriptions-change-payment-gateway.php#L340-L345). I believe it's done that way because that filter is intended for off-site payments (PayPal in particular), where the notice wouldn't actually display until after returning from that flow. _Is it acceptable to proceed with saving the payment method prior to authentication?_
  - If so, then we can consider showing another success message after authentication succeeds. That should be done in [this function](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/832aebe209ceeaf2b8f916dfec6ad5c0916bf3d1/includes/compat/class-wc-stripe-subs-compat.php#L675-L688) (though under what precise conditions is yet TBD).
  - If not, we might need to rework the flow a bit, maybe authenticating from the initial form without a reload (see https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1409)

- There is no error message shown if authentication fails – IIRC it was intended to show after redirect in the "Pay for order" screen case, but that isn't working either, so I think this can be raised outside the scope of the present issue.

- There is also currently no specific error displayed in the front-end if the SetupIntent creation fails (which happened to me somehow when using an old saved card that seems to have been attached to a different Stripe customer) – not sure if that's an issue.

- Instead of always creating a new SetupIntent, we might want to check for an existing SetupIntent and update it, as is done for PaymentIntents [here](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/832aebe209ceeaf2b8f916dfec6ad5c0916bf3d1/includes/class-wc-gateway-stripe.php#L610-L614)… but I'm not so sure. If a new choice is being made to change the card, I believe that should be a new SetupIntent by definition. If authentication for the same card is being retried, is there a strong reason to avoid creating a new SetupIntent?

- https://github.com/woocommerce/woocommerce-gateway-stripe/commit/0b8d05601923578ad953014820ee2f912be4e7a2 adds hooks for modifying the "Pay for order" screen on an _action_ instead of a _filter_ (so that the hooks can be removed on a subsequent action, in time for [this check](https://github.com/woocommerce/woocommerce-subscriptions/blob/45021ff801f141655c9ab56c0c7636ed89cfda77/includes/class-wc-subscriptions-change-payment-gateway.php#L197) which needs the available gateways). The gateways are still being filtered inside the function – asked about the duplication in https://github.com/woocommerce/woocommerce-gateway-stripe/commit/ec15b62f56bf84b7489ecf75db44954b9b69bd80#r36094893. Is there a reason it was done on a filter that I might have missed? cc @RadoslavGeorgiev 

(I tried to capture the reasons for things here and in commit messages but I'm not sure if I covered them all so of course feel free to ask.)

## Testing instructions

Use instructions from https://github.com/woocommerce/woocommerce-gateway-stripe/issues/1046 but **verify that modal authentication appears**, and that if it succeeds then **the renewal also succeeds**. (Just note that the "Edit Subscription" screen needs to be reloaded after the payment method is saved, or else the "Process renewal" action will actually update it back to the earlier source, being the value of the "Stripe Source ID" field in the admin form 🤔.)

(To change the payment method, navigate on the front-end to My account » Subscriptions » any active subscription and click "Change payment", or navigate in wp-admin to WooCommerce » Subscriptions » any active subscription and click "Customer change payment method page →".)

Can also test to make sure the ordinary Pay for Order flow (for regular pending orders) still works as before.

-------------------
